src/cargo/ops/tree/format/mod.rs

```diff
@@ -0,0 +1,108 @@
+                RawChunk::Argument("r") => Chunk::Repository,
+                RawChunk::Argument("f") => Chunk::Features,
+                RawChunk::Argument(a) => {
+                    return Err(anyhow!("unsupported pattern `{}`", a));
+                }
+                RawChunk::Error(err) => return Err(anyhow!("{}", err)),
+            };
+            chunks.push(chunk);
+        }
+
+        Ok(Pattern(chunks))
+    }
+
+    pub fn display<'a>(&'a self, graph: &'a Graph<'a>, node_index: usize) -> Display<'a> {
+        Display {
+            pattern: self,
+            graph,
+            node_index,
+        }
+    }
+}
+
+pub struct Display<'a> {
+    pattern: &'a Pattern,
+    graph: &'a Graph<'a>,
+    node_index: usize,
+}
+
+impl<'a> fmt::Display for Display<'a> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let node = self.graph.node(self.node_index);
+        match node {
+            Node::Package {
+                package_id,
+                features,
+                ..
+            } => {
+                let package = self.graph.package_for_id(*package_id);
+                for chunk in &self.pattern.0 {
+                    match *chunk {
+                        Chunk::Raw(ref s) => fmt.write_str(s)?,
+                        Chunk::Package => {
+                            write!(fmt, "{} v{}", package.name(), package.version())?;
+
+                            let source_id = package.package_id().source_id();
+                            if !source_id.is_default_registry() {
+                                write!(fmt, " ({})", source_id)?;
+                            }
+                        }
+                        Chunk::License => {
+                            if let Some(ref license) = package.manifest().metadata().license {
+                                write!(fmt, "{}", license)?;
+                            }
+                        }
+                        Chunk::Repository => {
+                            if let Some(ref repository) = package.manifest().metadata().repository {
+                                write!(fmt, "{}", repository)?;
+                            }
+                        }
+                        Chunk::Features => {
+                            write!(fmt, "{}", features.join(","))?;
+                        }
+                    }
+                }
+            }
+            Node::Feature { name, node_index } => {
+                let for_node = self.graph.node(*node_index);
+                match for_node {
+                    Node::Package { package_id, .. } => {
+                        write!(fmt, "{} feature \"{}\"", package_id.name(), name)?;
+                        if self.graph.is_cli_feature(self.node_index) {
+                            write!(fmt, " (command-line)")?;
+                        }
+                    }
+                    _ => panic!("unexpected feature node {:?}", for_node),
```

Comment:
Could this have a comment pointing to where it's otherwise validated this doesn't come up?